### PR TITLE
TYP: pd.eval, DataFrame.(eval, query) 

### PR DESCRIPTION
**Pull request by arw2019**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

xref https://github.com/pandas-dev/pandas/pull/38416#discussion_r546736072

### CODE REVIEW

File: pandas/core/computation/eval.py [hunk 1]
```diff
@@ -159,16 +160,16 @@ def _check_for_locals(expr: str, stack_level: int, parser: str):
 
 def eval(
     expr,
-    parser="pandas",
+    parser: str = "pandas",
     engine: Optional[str] = None,
-    truediv=no_default,
-    local_dict=None,
-    global_dict=None,
+    truediv: Optional[bool] = no_default,
```
Comment (jbrockmendel) [inline pandas/core/computation/eval.py:165]: Optional[bool] doesnt fit no_default

Comment (arw2019) [reply pandas/core/computation/eval.py]: thanks - will leave unannotated in this PR

File: pandas/core/computation/eval.py [hunk 2]
```diff
@@ -159,16 +160,16 @@ def _check_for_locals(expr: str, stack_level: int, parser: str):
 
 def eval(
     expr,
-    parser="pandas",
+    parser: str = "pandas",
```
Comment (jbrockmendel) [inline pandas/core/computation/eval.py:163]: @simonjayhawkins IIUC we can't use Literal yet.  what if we used enums as a stopgap

```
class ParserStr(enum):
    pandas = "pandas"
    numexpr = "numexpr"
```

that way we could start filling these in where appropriate and when we can use Literal swapping them out will be easy (wont require hunting them down)

File: pandas/core/frame.py [hunk 3]
```diff
@@ -3448,8 +3449,8 @@ def query(self, expr: str, inplace: bool = False, **kwargs):
 
         Returns
         -------
-        DataFrame or None
-            DataFrame resulting from the provided query expression or
+        ndarray, scalar, pandas object, or None
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3452]: "pandas object" means "Series or DataFrame" here?  could confuse with PandasObject

File: pandas/core/frame.py [hunk 4]
```diff
@@ -3448,8 +3449,8 @@ def query(self, expr: str, inplace: bool = False, **kwargs):
...
+            The result of evaluation of the provided query expression or
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3453]: "of evaluation of" -> "of evaluating"?

**Comment by simonjayhawkins** [comment]
@arw2019 can you address @jbrockmendel comments.

File: pandas/core/computation/eval.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/computation/eval.py:169]: i think we are now using Literal

**Comment by jbrockmendel** [comment]
Needs rebase and a couple of comments are outstanding, but otherwise I think this is pretty close.

**Comment by simonjayhawkins** [comment]
@jbrockmendel over to you.

**Review by MarcoGorelli — COMMENTED** [review]
There's no default arguments preceding `inplace` in either of these functions, so they should be straightforward to overload - this would obviate the need for the casts you've added (no objection to this being done in a follow-up PR though)

Also, I think @jbrockmendel 's commented on having a preference for using `# type: ignore` rather than `cast`, especially if it's for something which can be fixed later (e.g. here it can be fixed with overloads)

**Comment by simonjayhawkins** [comment]
@MarcoGorelli feel free to push or otherwise close. I was going to close as stale but @jbrockmendel commented it was pretty close. https://github.com/pandas-dev/pandas/pull/38625#issuecomment-808916410

File: pandas/_typing.py [hunk 5]
```diff
@@ -160,6 +160,9 @@
 
 PythonFuncType = Callable[[Any], Any]
 
+IndexableEvalResult = Union[ArrayLike, FrameOrSeriesUnion]
+EvalResult = Union[IndexableEvalResult, Scalar]
```
Comment (jbrockmendel) [inline pandas/_typing.py:164]: is this meaningfully distinct from Any?

**Comment by simonjayhawkins** [comment]
closing as stale.

**Pull request closed without merging**